enhancements/machine-config/mco-network-configuration.md

```diff
@@ -0,0 +1,253 @@
+  - TBD
+creation-date: 2020-06-25
+last-updated: 2020-06-25
+status: provisional
+---
+
+# Extend MCO to support declarative host network configuration
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The ability to configure additional host networking parameters is paramount on some platforms, especially baremetal. Baremetal nodes usually have many interfaces that need to be dynamically configured to access other, critical datacenter resources. Doing so with MachineConfig would incur reboots that are expensive in time on the baremetal platform. The ability to configure host networking parameters without rebooting is desirable. It is also desirable to rollback automatically on failure, as opposed to failing on reboot.
+
+## Motivation
+
+With hybrid clouds, node-networking setup is becoming even more challenging.
+Different payloads have different networking requirements, and not everything
+can be satisfied as overlays on top of the main interface of the node (e.g.
+SR-IOV, L2, other L2).
+
+The [Container Network Interface](https://github.com/containernetworking/cni) (CNI) standard enables different solutions for connecting networks on the node with pods. Some of them are [part of the standard](https://github.com/containernetworking/plugins), and there are others that extend support for [Open vSwitch bridges](https://github.com/kubevirt/ovs-cni), [SR-IOV](https://github.com/hustcat/sriov-cni), and more...
+
+However, in all of these cases, the node must have the networks setup before the pod is scheduled. Setting up the networks in a dynamic and heterogenous cluster, with dynamic networking requirements, is a challenge by itself - and this is what this project is addressing.
+
+The node network configuration shall remain in the MachineConfigOperator workflow, but be declarative.
+
+### Goals
+
+* Enable an admin to modify node networking configuration without rebooting (Day 2)
+* Enable cluster administrators to modify node networking after cluster installation (day 2).
+* Integration with MachineConfig node configuration for consistency, familiarity and roll out functionality.
+* Rebootless configuration without deferring errors to a later reboot
+* Rolling back cluster breaking configuration mistakes.
+* 
+### Non-Goals
+
+* Replace SRIOV operator
+
+## Proposal
+
+#### Option A
+
+The node networking configuration shall be driven by MCO (and by extension kubernetes-nmstate) and executed by
+[nmstate](https://nmstate.github.io/).
+
+Do a KubeletConfig like integration:
+
+* NodeNetworkState will be introduced. This allows the user to see the network state of a node in the same declarative language that can be used to configure it.
```

Comment:
In some cases, there are needs to attach the host network devices in the container for application workload, for example, in a OSP VM environment, the virtio-net device may be attached to the pod container via host-device CNI, how would the NodeNetworkState handles this situation where the device disappear from the host network namespace?

Reply:
In some other cases, the host device may need to be provisioned with new profile/firmware or bitstream (for FPGA devices), will this functionality be considered in the proposal?

Reply:
Please take a new look at the proposal. We have introduced another alternative that we are now seeking,